File: Tps/Tp_2026/TP02/Questao01/Restaurante.java
import java.util.Scanner;

class Restaurante
{
    private int id;
    private String nome;
    private String cidade;
    private int capacidade;
    private double avaliacao;
    private String[] tiposCozinha;
    private String faixaPreco;
    private Hora horarioAbertura;
    private Hora horarioFechamento;
    private Data dataAbertura;
    private boolean aberto;
    

    /**
     * @author Gabriel Ferreira Pereira
     * @param id, nome, cidade, capacidade, avaliacao, tiposCozinha, faixaPreco, horarioAbertura, 
     *          horarioFechamento, dataAbertura, aberto do restaurante
     * @reason Construtor da classe Restaurante
     */
    public Restaurante( int id, String nome, String cidade, int capacidade, double avaliacao, 
        String[] tiposCozinha, String faixaPreco, Hora horarioAbertura, Hora horarioFechamento, Data dataAbertura, boolean aberto ) 
    {
        this.id = id;
        this.nome = nome;
        this.cidade = cidade;
        this.capacidade = capacidade;
        this.avaliacao = avaliacao;
        this.tiposCozinha = tiposCozinha;
        this.faixaPreco = faixaPreco;
        this.horarioAbertura = horarioAbertura;
        this.horarioFechamento = horarioFechamento;
        this.dataAbertura = dataAbertura;
        this.aberto = aberto;
    }

    // Getters

    public int getID ()
    {
        return this.id;
    }

    public String getNome ()
    {
        return this.nome;
    }

    public String getCidade ()
    {
        return this.cidade;
    }

    public int getCapacidade ()
    {
        return this.capacidade;
    }

    public double getAvaliacao ()
    {
        return this.avaliacao;
    }

    public String[] getTiposCozinha ()
    {
        return this.tiposCozinha;
    }

    public String getFaixaPreco ()
    {
        return this.faixaPreco;
    }

    public Hora getHorarioAbertura ()
    {
        return this.horarioAbertura;
    }

    public Hora getHorarioFechamento ()
    {
        return this.horarioFechamento;
    }

    public Data getDataAbertura ()
    {
        return this.dataAbertura;
    }

    public boolean getAberto ()
    {
        return this.aberto;
    }
    
    /**
 * @author Gabriel Ferreira Pereira
 * @param String s
 * @reason Converte uma String para double manualmente
 * @return numero real
 */
public static double parseDouble(String s)
{
    int inteiro = 0;
    int decimal = 0;
    boolean ponto = false;

    for (int i = 0; i < s.length(); i++)
    {
        char c = s.charAt(i);
        if (c == '.')
            ponto = true;
        else if (!ponto)
            inteiro = inteiro * 10 + (c - '0');
        else
            decimal = decimal * 10 + (c - '0');
    }

    return inteiro + decimal / 10.0;
}

    /**
     * @author Gabriel Ferreira Pereira
     * @param path caminho do arquivo CSV
     * @reason Le uma linha do CSV e cria um objeto Restaurante
     * @return Restaurante criado
     */
    public static Restaurante ler( String linha ) throws Exception
    {
        Scanner sc = new Scanner(linha);
        int id = sc.nextInt();
        String nome = sc.next();
        String cidade = sc.next();
        int capacidade = sc.nextInt();
        double avaliacao = parseDouble(sc.next());

        String cozinhaStr = sc.next();
        Scanner scCozinha = new Scanner(cozinhaStr);
        scCozinha.useDelimiter(";");
        String tipo1 = scCozinha.next();
        String tipo2 = scCozinha.next();
        scCozinha.close();
        String[] tiposCozinha = new String[]{tipo1, tipo2};

        String faixaPreco = sc.next();

        String horario = sc.next();
        Scanner scHorario = new Scanner(horario);
        scHorario.useDelimiter("-");
        Hora horarioAbertura   = Hora.parseHora(scHorario.next());
        Hora horarioFechamento = Hora.parseHora(scHorario.next());
        scHorario.close();

        Data dataAbertura = Data.parseData(sc.next());

        String abertoStr = sc.next();
        boolean aberto = abertoStr.charAt(0) == 't';

        return new Restaurante(id, nome, cidade, capacidade, avaliacao, tiposCozinha, faixaPreco, horarioAbertura, horarioFechamento, dataAbertura, aberto);
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @reason Retorna o restaurante formatado como String
     * @return String com o restaurante formatado
     */
    public String formatar()
    {
        String tipos = "[" + tiposCozinha[0] + "," + tiposCozinha[1] + "]";
        String horario = horarioAbertura.formatarHora() + "-" + horarioFechamento.formatarHora();
        String data = dataAbertura.formatarData();
    
        return "[" + id + " ## " + nome + " ## " + cidade + " ## " + capacidade + " ## " + 
            avaliacao + " ## " + tipos + " ## " + faixaPreco + " ## " + 
            horario + " ## " + data + " ## " + aberto + "]";
    }

}

File: Tps/Tp_2026/TP02/Questao03/Restaurante.c
#include "Restaurante.h"
#include <stdio.h>
#include <string.h>

/**
 * @author Gabriel Ferreira Pereira
 * @param linha, pos, campo
 * @reason Le um campo da linha do CSV ate encontrar virgula ou fim de linha
 * @return nova posicao na linha apos o campo lido
 */
int lerCampo( char *linha, int pos, char *campo )
{
    int i = 0;
    
    while ( linha[pos] != ',' && linha[pos] != '\0' && linha[pos] != '\n' )
    {
        campo[i] = linha[pos];
        i++;
        pos++;
    }

    campo[i] = '\0';
    pos++;

    return pos;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param linha String da linha do CSV
 * @reason Le uma linha do CSV e cria um objeto Restaurante
 * @return Restaurante criado
 */
Restaurante ler_restaurante( char *linha )
{
    Restaurante r;
    char campo[100];
    char horarioAbertura[10];
    char horarioFechamento[10];
    char dataStr[20];
    char abertoStr[10];
    char cozinha[100];
    int pos = 0;
    int j = 0, k = 0;

    // id
    pos = lerCampo(linha, pos, campo);
    sscanf(campo, "%d", &r.id);

    // nome
    pos = lerCampo(linha, pos, r.nome);

    // cidade
    pos = lerCampo(linha, pos, r.cidade);

    // capacidade
    pos = lerCampo(linha, pos, campo);
    sscanf(campo, "%d", &r.capacidade);

    // avaliacao
    pos = lerCampo(linha, pos, campo);
    sscanf(campo, "%lf", &r.avaliacao);

    // cozinha
    pos = lerCampo(linha, pos, cozinha);
    while (cozinha[j] != ';')
    {
        r.tipo1[k++] = cozinha[j++];
    }
    
    r.tipo1[k] = '\0';
    j++; k = 0;
    
    while (cozinha[j] != '\0')
    {
        r.tipo2[k++] = cozinha[j++];
    }
    
    r.tipo2[k] = '\0';

    // faixa preco
    pos = lerCampo(linha, pos, r.faixaPreco);

    // horario
    pos = lerCampo(linha, pos, campo);
    j = 0; k = 0;
    
    while (campo[j] != '-')
    {
        horarioAbertura[k++] = campo[j++];
    }
    
    horarioAbertura[k] = '\0';
    
    j++; k = 0;
    while (campo[j] != '\0')
    {
        horarioFechamento[k++] = campo[j++];
    }
    
    horarioFechamento[k] = '\0';
    r.horarioAbertura = parse_hora(horarioAbertura);
    r.horarioFechamento = parse_hora(horarioFechamento);

    // data
    pos = lerCampo(linha, pos, dataStr);
    r.dataAbertura = parse_data(dataStr);

    // aberto
    pos = lerCampo(linha, pos, abertoStr);

    if (abertoStr[0] == 't')
    {
        r.aberto = 1;
    }
    
    else
    {
        r.aberto = 0;
    }

    return r;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param r, saidaLinha
 * @reason Retorna o restaurante formatado como String
 */
void formatar_restaurante(Restaurante *r, char *saidaLinha)
{
    char dataStr[20];
    char horarioAbertura[10];
    char horarioFechamento[10];
    char abertoStr[10];

    formatar_data(&r->dataAbertura, dataStr);
    formatar_hora(&r->horarioAbertura, horarioAbertura);
    formatar_hora(&r->horarioFechamento, horarioFechamento);

    if (r->aberto == 1)
    {
        sprintf(abertoStr, "true");
    }
    
    else
    {
        sprintf(abertoStr, "false");
    }

    sprintf(saidaLinha, "[%d ## %s ## %s ## %d ## %.1f ## [%s,%s] ## %s ## %s-%s ## %s ## %s]",
        r->id, r->nome, r->cidade, r->capacidade, r->avaliacao,
        r->tipo1, r->tipo2, r->faixaPreco,
        horarioAbertura, horarioFechamento,
        dataStr, abertoStr);
}